[extension][ecsobserver] Init ECS Observer README

pingleig opened the pull request:
**Description:** <Describe what has changed.>

Follow up on #1920, this is the (slightly updated) doc for implementing service discovery (of scrape targets) for Prometheus receiver on Amazon Elastic Container (ECS). It's mainly based on the existing configuration & implementation we have in 
[amazon-cloudwatch-agent](https://github.com/aws/amazon-cloudwatch-agent/tree/master/internal/ecsservicediscovery)

There are three types of filter, we fetch all the tasks periodically and there is no server side filtering or list watch API.

- service i.e. deployment name
- task definition i.e. manifest
- docker label

The discovered targets are consumed in two ways, write as a static file for file sd, or use the receiver creator framework to create new receiver at runtime, based on #1395 the latter seems to have performance problem (especially when a single collector instance is used for scraping a large cluster w/o sharding)

There are some open items

- [x] emit custom metric, `discoverd_targets`, `api_error_counts` etc. Should be able to use [obsreport](https://github.com/open-telemetry/opentelemetry-collector/blob/main/obsreport/doc.go)
- [x] In cloudwatch agent, we were attaching ecs labels directly i.e. `InstanceId` instead of generating meta label like `__meta__ecs_task_arn__`, `__meta_ec2_instance_id__` and have user configuring relabel config. The `__meta__` approaches is the one used by all the in tree discovery implementations (k8s, consul etc.) (Chose to use `__meta_ecs_` prefix)

**Link to tracking Issue:**  #1395 

**Testing:** <Describe what testing was performed and which tests were added.>

N/A just the doc

**Documentation:** <Describe the documentation added.>

A new README under `extensions/observer/ecsobserver`

pingleig committed:
ecsobserver: Init README

jrcamp commented on extension/observer/ecsobserver/README.md:
```diff
@@ -0,0 +1,330 @@
+      job_name: guilty-spark
+  task_definitions:
+    - arn_pattern: '*memcached.*'
+    - arn_pattern: '^proxy-.*$`
```
```suggestion
    - arn_pattern: '^proxy-.*$'
```

pingleig replied:
done

jrcamp reviewed (commented):
@pingleig are you still working on changes in response to original PR? Just wanted to make sure not blocked on me.

anuraaga commented on extension/observer/ecsobserver/README.md:
```diff
@@ -0,0 +1,330 @@
+
+**Status: beta**
+
+The `ecsobserver` uses the ECS/EC2 API to discover prometheus scrape target from all running tasks and filter them based
```
```suggestion
The `ecsobserver` uses the ECS/EC2 API to discover prometheus scrape targets from all running tasks and filter them based
```

pingleig replied:
doine

mxiamxia commented on extension/observer/ecsobserver/README.md:
```diff
@@ -0,0 +1,330 @@
+| cluster_name     | Mandatory | target ECS cluster name for service discovery                                                    |
+| cluster_region   | Mandatory | target ECS cluster's AWS region name                                                             |
+| refresh_interval | Optional  | how often to look for changes in endpoints (default: 10s)                                        |
+| result_file      | Optional  | path of YAML file for the scrape target results. When enabled, the observer always returns empty |
```
Can `result_file` field be mandatory? Otherwise, what's purpose of observer?

pingleig replied:
Changed to mandatory for now,  it's not mandatory in the original PR because receiver creator is supported and the extension just returns the endpoint to observer listener. To reduce the size of PRs I plan to only implement the result file logic.

mxiamxia commented on extension/observer/ecsobserver/README.md:
```diff
@@ -0,0 +1,330 @@
+
+## Known issues
+
+- There is no list watch API in ECS (unlike k8s), and we fetch ALL the tasks and filter it locally. If the poll interval
```
can we add jitter on ECS API callings for the collectors in the same cluster?

pingleig replied:
I think jitter may not be necessary unless all the collectors (w/ ecs discovery) are started at the same time. We don't deploy it as a daemon service, so chance of this scenario is low. We do have cache on some API results (ECS TaskDefinition, EC2 Instance) so we should be good. Though I plan to emit some metrics/log so we can know it when some users are getting to much API errors.

pingleig committed:
ecsobserver: Update output format to use __meta_ecs_ prefix

pingleig commented:
@jrcamp Sorry for the delay, I wasn't blocked on the comments in original PR. I was has having some internal discussion on the output format. The conclusion is we will be using `__meta__ecs_` prefix like other in tree service discovery implementations.

tigrannajaryan commented:
```
./extension/observer/ecsobserver/README.md:21:4: "sevice" is a misspelling of "service"
./extension/observer/ecsobserver/README.md:399:3: "discoverd" is a misspelling of "discovered"
./extension/observer/ecsobserver/README.md:400:3: "discoverd" is a misspelling of "discovered"
./extension/observer/ecsobserver/README.md:401:74: "discoverd" is a misspelling of "discovered"
make: *** [Makefile.Common:119: misspell] Error 2
make: *** Waiting for unfinished jobs....
```

bogdandrutu commented:
ping @jrcamp

pingleig committed:
ecsobserver: Fix typo

- `make misspell`
- Fix order of filter types, should be service, task defintion, docker
  label

jrcamp commented:
Looks good, thanks for the thorough README!

pingleig commented:
Can anyone with merge permission merge this PR? I think I don't need to do rebase on my fork because it's a README in a new folder and should not have any conflict.

pingleig commented:
ping @jrcamp @bogdandrutu
